### DISCUSSION TITLE
ESP32+ILI9341 resets on tft.begin()

### DISCUSSION THREAD

**Discussion by HeikoH** [Bodmer/TFT_eSPI#1571] [Q&A - General]
The moment I call `tft.begin()` my ESP32 board (LOLIN32 Lite) with a ILI9341 display resets. Any idea what could be wrong? 

```c
#define USER_SETUP_LOADED
#define ILI9341_DRIVER
#define TFT_DC  16
#define TFT_RST 17
#define SPI_FREQUENCY       20000000
#define SPI_READ_FREQUENCY  10000000

#include <TFT_eSPI.h>

#define BACKLIGHT 4

TFT_eSPI tft = TFT_eSPI();

void setup() {
  Serial.begin(115200);
  Serial.println("");
  Serial.println("---");

  pinMode(BACKLIGHT, OUTPUT);
  digitalWrite(BACKLIGHT, HIGH);

  tft.init();
}

void loop() {
}
```

This results in a continues loop of resets:
```
---
ets Jun  8 2016 00:22:57

rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:12804
ho 0 tail 12 room 4
load:0x40080400,len:3032
entry 0x400805e4

---
ets Jun  8 2016 00:22:57

rst:0x8 (TG1WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:12804
ho 0 tail 12 room 4
load:0x40080400,len:3032
entry 0x400805e4

etc.... etc... etc..
```

note; a similar program using a different TFT library works fine, same device:
```c
#include "Adafruit_ILI9341.h"

#define TFT_CS -1
#define TFT_DC 16
#define TFT_RST 17
#define BACKLIGHT 4

Adafruit_ILI9341 tft = Adafruit_ILI9341(TFT_CS, TFT_DC, TFT_RST);
void setup() {
  Serial.begin(115200);
  Serial.println("");
  Serial.println("---");

  pinMode(BACKLIGHT, OUTPUT);
  digitalWrite(BACKLIGHT, HIGH);

  tft.begin();
  tft.fillScreen(ILI9341_RED);
}

void loop() {
}
```

**Comment by Bodmer** [comment]
With TFT_eSPI you CANNOT define the pins in the sketch, you must edit the library setup file when using the Arduino IDE. In PlatformIO you can use the platformio.ini file to define the pins.

I expect that illegal pins are being declared in the setup file, this then casues the ESP32 to crash..

**Reply by shan-mcarthur** [reply]
Isn't the point that we can define the pins in the sketch (before the include of TFT_eSPI.h) and also define USER_SETUP_LOADED?  These macros are defined in the libraries' User_Setup.h file which when detected in User_Setup_Select.h skips including that file.  I reviewed the full user setup file and noticed that I didn't include a few macros, including SMOOTH_FONT and some SPI frequencies.  When I added those, it stopped rebooting.  I don't think rebooting is the best answer here, but one small suggestion is to make sure that the required macros are called out that if left out would cause constant reboots.

**Comment by HeikoH** [comment]
@Bodmer - defining stuff like in my snippet should be perfectly fine. But, I found the issue... I was missing `#define SMOOTH_FONT`, seems not defining it (in the sketch or in `User_Setup.h`) causes the reboot. For posterity a working example:
```c
#define USER_SETUP_LOADED
#define ILI9341_DRIVER
#define TFT_DC  16
#define TFT_RST 17
#define SMOOTH_FONT

#include <TFT_eSPI.h>

#define BACKLIGHT 4

TFT_eSPI tft = TFT_eSPI();

void setup() {
  pinMode(BACKLIGHT, OUTPUT);
  digitalWrite(BACKLIGHT, HIGH);

  tft.init();
}

void loop() {
    tft.fillScreen(ILI9341_RED);
    delay(100);
    tft.fillScreen(ILI9341_GREEN);
    delay(100);
}
```

**Reply by hi-one-gg** [reply]
So you mean that declaring `#define SMOOTH_FONT` will stop the infinite reboot?

I encountered the same error. I have tried every possible method for four days but have failed. The only solution is to downgrade the board version to 2.0.14. Apart from that, there is no solution. Even the creators have not provided a clear answer.
